src/mscorlib/src/System/Span.cs

```diff
@@ -164,5 +167,145 @@ public bool Equals(Span<T> other)
+        {
+            SpanContracts.RequresNoReferences<T>();
+            SpanContracts.RequiresNonNegative(elementsCount);
+            SpanContracts.RequiresInInclusiveRange(_length, elementsCount);
+
+            SpanHelper.Memmove<T>((byte*)_rawPointer, (byte*)source, elementsCount);
+        }
+    }
+
+    internal static class SpanContracts
+    {
+        internal static void RequiresIndexInRange(int index, int length) {
+            if ((uint)index >= (uint)length)
+                ThrowHelper.ThrowIndexOutOfRangeException();
+        }
+
+        internal static void RequiresInRange(int start, int length) {
+            if ((uint)start >= (uint)length)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+        }
+
+        public static void RequiresInInclusiveRange(int start, int length) {
+            if ((uint)start > (uint)length)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+        }
+
+        public static void RequiresNonNegative(int number) {
+            if (number < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+        }
+
+        internal static void RequiresNonNullArray<T>(T[] array) {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+        }
+
+        internal static void RequiresNoCovariance<T>(T[] array) {
+            if (default(T) == null) { // Arrays of valuetypes are never covariant
+                if (array.GetType() != typeof(T[]))
+                    ThrowHelper.ThrowArrayTypeMismatchException();
+            }
+        }
+
+        /// <summary>
+        /// throws for reference types and value types with references, they must not be put to unmanaged memory
+        /// </summary>
+        internal static void RequresNoReferences<T>() {
+            if (JitHelpers.ContainsReferences<T>())
+                ThrowHelper.ThrowInvalidTypeForUnmanagedMemory(typeof(T));
+        }
+    }
+
+    internal static class SpanHelper
+    {
+        internal static void CopyTo<T>(ref IntPtr destination, ref IntPtr source, int elementsCount)
+        {
+            if (!JitHelpers.ContainsReferences<T>())
+            {
+                unsafe
+                {
+                    Memmove<T>((byte*)destination, (byte*)source, elementsCount);
+                }
+            }
+            else
+            {
+                ref T dest = ref JitHelpers.GetByRef<T>(ref destination);
+                ref T src = ref JitHelpers.GetByRef<T>(ref source);
+
+                for (int i = 0; i < elementsCount; i++)
+                {
+                    JitHelpers.AddByRef(ref dest, i) = JitHelpers.AddByRef(ref src, i);
+                }
+            }
+        }
+
+        internal static unsafe void Memmove<T>(byte* destination, byte* source, int elementsCount)
+        {
+#if BIT64
+            Buffer.Memmove(destination, source, (ulong)(elementsCount * (uint)JitHelpers.SizeOf<T>()));
```

Comment:
Could you please cast the SizeOf to ulong to make it easy to see that the multiplication is done as 64-bit?